lib/thrift/binary/framed/client.ex

```diff
@@ -291,10 +305,59 @@ defmodule Thrift.Binary.Framed.Client do
     end
   end
 
+  @impl Connection
+  def handle_info({:tcp, sock, _data}, %{reconnect: true, sock: {:gen_tcp, sock}} = s) do
+    {:disconnect, :reconnect, s}
+  end
+
+  def handle_info({:tcp_closed, sock}, %{reconnect: true, sock: {:gen_tcp, sock}} = s) do
+    {:disconnect, :reconnect, s}
+  end
+
+  def handle_info({:tcp_error, sock, _error}, %{reconnect: true, sock: {:gen_tcp, sock}} = s) do
+    {:disconnect, :reconnect, s}
+  end
+
+  def handle_info({:ssl, sock, _data}, %{reconnect: true, sock: {:ssl, sock}} = s) do
+    {:disconnect, :reconnect, s}
+  end
+
+  def handle_info({:ssl_closed, sock}, %{reconnect: true, sock: {:ssl, sock}} = s) do
+    {:disconnect, :reconnect, s}
+  end
+
+  def handle_info({:ssl_error, sock, _error}, %{reconnect: true, sock: {:ssl, sock}} = s) do
+    {:disconnect, :reconnect, s}
```

Comment:
I think we should always disconnect and make the decision on reconnect there. If `reconnect: false` we want to disconnect too.

Reply:
In this case many tests fails. If I disconnect on any TCP or SSL message with `reconnect: false` following output produced by `mix test`:

```
19:52 $ mix test
Compiling 1 file (.ex)
Excluding tags: [pending: true]

............................................................................................................................................

  1) test clients exit if they try to use a closed client (Thrift.Generator.ServiceTest)
     test/thrift/generator/service_test.exs:387
     ** (exit) exited in: :gen_server.call(#PID<0.1755.0>, {:call, "friend_ids_of", [<<10, 0, 1, 0, 0, 0, 0, 0, 0, 4, 210>> | <<0>>], []}, 5000)
         ** (EXIT) {:error, :closed}
     code: assert Client.friend_ids_of(client, 1234) == {:error, :closed}
     stacktrace:
       (stdlib) gen_server.erl:223: :gen_server.call/3
       (thrift) lib/thrift/binary/framed/client.ex:221: Thrift.Binary.Framed.Client.call/5
       test/thrift/generator/service_test.exs:392: (test)

     The following output was logged:
     
     19:52:36.114 [error] Connection closed
     
     19:52:36.115 [error] GenServer #PID<0.1755.0> terminating
     ** (stop) {:error, :closed}
     Last message: {:tcp_closed, #Port<0.19>}
     State: nil
     
..............

  2) test clients exit on connection timeout (Thrift.Generator.ServiceTest)
     test/thrift/generator/service_test.exs:373
     ** (exit) exited in: :gen_server.call(#PID<0.1958.0>, {:call, "friend_ids_of", [<<10, 0, 1, 0, 0, 0, 0, 0, 0, 4, 210>> | <<0>>], []}, 5000)
         ** (EXIT) no process: the process is not alive or there's no process currently associated with the given name, possibly because its application isn't started
     code: assert {:error, closed} = Client.friend_ids_of(client, 1234)
     stacktrace:
       (stdlib) gen_server.erl:223: :gen_server.call/3
       (thrift) lib/thrift/binary/framed/client.ex:221: Thrift.Binary.Framed.Client.call/5
       test/thrift/generator/service_test.exs:382: (test)

     The following output was logged:
     
     19:52:36.155 [error] Connection closed
     
     19:52:36.155 [error] GenServer #PID<0.1958.0> terminating
     ** (stop) {:error, :closed}
     Last message: {:tcp_closed, #Port<0.59>}
     State: nil
     
..

  3) test it handles void oneway functions (Thrift.Generator.ServiceTest)
     test/thrift/generator/service_test.exs:309
     ** (EXIT from #PID<0.1990.0>) {:error, :closed}

     The following output was logged:
     
     19:52:36.188 [error] Connection closed
     
     19:52:36.188 [error] GenServer #PID<0.1996.0> terminating
     ** (stop) {:error, :closed}
     Last message: {:tcp, #Port<0.68>, <<0>>}
     State: nil
     
     19:52:36.189 [error] GenServer #PID<0.1991.0> terminating
     ** (stop) {:error, :closed}
     Last message: {:EXIT, #PID<0.1990.0>, {:error, :closed}}
     State: {:state, {#PID<0.1991.0>, Supervisor.Default}, :one_for_one, {[ranch_listener_sup: Thrift.Generator.ServiceTest.ServerSpy], %{{:ranch_listener_sup, Thrift.Generator.ServiceTest.ServerSpy} => {:child, #PID<0.1992.0>, {:ranch_listener_sup, Thrift.Generator.ServiceTest.ServerSpy}, {:ranch_listener_sup, :start_link, [Thrift.Generator.ServiceTest.ServerSpy, :ranch_tcp, %{num_acceptors: 1, socket_opts: [port: 0]}, Thrift.Binary.Framed.ProtocolHandler, {Thrift.Generator.ServiceTest.SimpleService.Binary.Framed.Server, Thrift.Generator.ServiceTest.ServerSpy, [], []}]}, :permanent, :infinity, :supervisor, [:ranch_listener_sup]}}}, :undefined, 0, 5, [], 0, Supervisor.Default, {:ok, {%{intensity: 0, period: 5, strategy: :one_for_one}, [{{:ranch_listener_sup, Thrift.Generator.ServiceTest.ServerSpy}, {:ranch_listener_sup, :start_link, [Thrift.Generator.ServiceTest.ServerSpy, :ranch_tcp, %{num_acceptors: 1, socket_opts: [port: 0]}, Thrift.Binary.Framed.ProtocolHandler, {Thrift.Generator.ServiceTest.SimpleService.Binary.Framed.Server, Thrift.Generator.ServiceTest.ServerSpy, [], []}]}, :permanent, :infinity, :supervisor, [:ranch_listener_sup]}]}}}
     
.................

  4) test it can handle oneway messages (Servers.Binary.Framed.IntegrationTest)
     test/thrift/binary/framed/server_test.exs:158
     ** (EXIT from #PID<0.2192.0>) {:error, :closed}

     The following output was logged:
     
     19:52:36.324 [error] Connection closed
     
     19:52:36.324 [error] GenServer :"test it can handle oneway messages" terminating
     ** (stop) {:error, :closed}
     Last message: {:tcp, #Port<0.114>, <<0>>}
     State: nil
     
..................................................

Finished in 6.3 seconds
227 tests, 4 failures

Randomized with seed 350387
```

Reply:
I'm not sure - should I correct this test cases?

Reply:
```

Reply:
Hi, any comments about my last question?

Reply:
We need to think about we handle these cases, and handle them.

For example we could do

```elixir
try do
  GenServer.call(...)
catch
  :exit, {{:error, _} = error, _} ->
    error
  :exit, {:noproc, _} ->
    {:error, :closed}
end
```

Reply:
My opinion is not to handle these cases and let users of library to handle it. But in this case some test should be rewritten